File: cat/cat.c
// cat.c 
// Author: jcolours 
// Date: 2026-06-12
// Description: Simple remake of the basic cat unix tool using C, 
// C libraries, and C sys calls to the kernel

#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>

int main(int argc, char *argv[])
{
  if (argc > 1)
  {
    for (int i=1; i < argc; i++)
    {
      int fd;

      if ((fd = open(argv[i], O_RDONLY)) == -1)
      {
        perror("\nopen");
        exit(EXIT_FAILURE);
      }

      char buf[4096];

      size_t size;
      if ((size = read(fd, buf, sizeof(buf))) == -1)
      {
        perror("\nread");
        exit(EXIT_FAILURE);
      }

      buf[size] = '\0';

      printf("%s", buf);

      close(fd);
    }
  }
  else 
  {
    int fd = 0;

    char buf[4096];

    size_t size = 1;

    while (size > 0) // reason: read returns 0 on EOF and -1 on error
    {
      if ((size = read(fd, buf, sizeof(buf)) == -1)
      {
        perror("\nread");
        exit(EXIT_FAILURE);
      }
  
      buf[size] = '\0';

      printf("%s", buf);
    }

    close(fd);
  }

  return EXIT_SUCCESS;
}

File: unix-tools-go/cat_go/cat.go
package main

import (
	"fmt"
	"io"
	"log"
	"os"
)

func main() {
	fmt.Printf("\n")

	if len(os.Args) <= 1 {

		buf := make([]byte, 4096)

		for {
			buf_size, err := os.Stdin.Read(buf)

			if err == io.EOF { break }

			for i := 0; i < buf_size; i++ { fmt.Printf("%c", buf[i]) }
		}

	} else {
		for i := 1; i < len(os.Args); i++ {

			file, err := os.Open(os.Args[i])

			if err != nil { log.Fatal(err) }

			buf := make([]byte, 4096)

			for {
				buf_size, err := file.Read(buf)

				if err == io.EOF { break }

				for j := 0; j < buf_size; j++ { fmt.Printf("%c", buf[j]) }
			}

			fmt.Println()

			file.Close()
		}
	}

	os.Exit(0)
}
